kotlin-formal/KotlinParser.g4

```diff
@@ -0,0 +1,892 @@
+    (NL* typeConstraints)?
+    (NL* classBody | NL* enumClassBody)?
+    ;
+
+primaryConstructor
+    : (modifiers? 'constructor' NL*)? classParameters
+    ;
+
+classParameters
+    : '(' NL* (classParameter (NL* ',' NL* classParameter)*)? NL* ')'
+    ;
+
+classParameter
+    : modifiers? ('val' | 'var')? NL* simpleIdentifier ':' NL* type (NL* '=' NL* expression)?
+    ;
+
+delegationSpecifiers
+    : annotatedDelegationSpecifier (NL* ',' NL* annotatedDelegationSpecifier)*
+    ;
+
+annotatedDelegationSpecifier
+    : annotation* NL* delegationSpecifier
+    ;
+
+delegationSpecifier
+    : constructorInvocation
+    | explicitDelegation
+    | userType
+    | functionType
+    ;
+
+constructorInvocation
+    : userType valueArguments
+    ;
+
+explicitDelegation
+    : (userType | functionType) NL* 'by' NL* expression
+    ;
+
+classBody
+    : '{' NL* classMemberDeclarations NL* '}'
+    ;
+
+classMemberDeclarations
+    : (classMemberDeclaration semis?)*
+    ;
+
+classMemberDeclaration
+    : declaration
+    | companionObject
+    | anonymousInitializer
+    | secondaryConstructor
+    ;
+
+anonymousInitializer
+    : 'init' NL* block
+    ;
+
+secondaryConstructor
+    : modifiers? 'constructor' NL* functionValueParameters (NL* ':' NL* constructorDelegationCall)? NL* block?
+    ;
+
+constructorDelegationCall
+    : 'this' NL* valueArguments
+    | 'super' NL* valueArguments
+    ;
+
+enumClassBody
+    : '{' NL* enumEntries? (NL* ';' NL* classMemberDeclarations)? NL* '}'
+    ;
+
+enumEntries
+    : enumEntry (NL* ',' NL* enumEntry)* NL* ','?
+    ;
+
+enumEntry
+    : (modifiers NL*)? simpleIdentifier (NL* valueArguments)? (NL* classBody)?
+    ;
+
+functionDeclaration
+    : modifiers?
+    'fun' (NL* typeParameters)? (NL* receiverType NL* '.')? NL* simpleIdentifier
+    NL* functionValueParameters
+    (NL* ':' NL* type)?
+    (NL* typeConstraints)?
+    (NL* functionBody)?
+    ;
+
+functionValueParameters
+    : '(' NL* (functionValueParameter (NL* ',' NL* functionValueParameter)*)? NL* ')'
+    ;
+
+functionValueParameter
+    : modifiers? parameter (NL* '=' NL* expression)?
+    ;
+
+parameter
+    : simpleIdentifier NL* ':' NL* type
+    ;
+
+setterParameter
+    : simpleIdentifier NL* (':' NL* type)?
+    ;
+
+functionBody
+    : block
+    | '=' NL* expression
+    ;
+
+objectDeclaration
+    : modifiers? 'object'
+    NL* simpleIdentifier
+    (NL* ':' NL* delegationSpecifiers)?
+    (NL* classBody)?
+    ;
+
+companionObject
+    : modifiers? 'companion' NL* 'object'
+    (NL* simpleIdentifier)?
+    (NL* ':' NL* delegationSpecifiers)?
+    (NL* classBody)?
+    ;
+
+propertyDeclaration
+    : modifiers? ('val' | 'var')
+    (NL* typeParameters)?
+    (NL* receiverType NL* '.')?
+    (NL* (multiVariableDeclaration | variableDeclaration))
+    (NL* typeConstraints)?
+    (NL* ('=' NL* expression | propertyDelegate))?
+    (NL+ ';')? NL* (getter? (NL* semi? setter)? | setter? (NL* semi? getter)?)
```

Comment:
This does look complicated, however what do you think of 

```
(semi getter (NL* semi? setter)? | semi setter (NL* semi? getter)?)?
```
?

I guess it is pretty much the same grammar as you write, however it does not cause ambiguity for the case when no getter/setter is present.
If both getter and setter are absent, then "closing `(NL+ ';')? NL*` causes ambiguity with `semis` in-between `statements`.

Reply:
This semicolon mess is trying to replicate the compiler behaviour, which is quite illogical here. The difference between `';'` and `semi` here is important, as far as I remember.
Your variant seems legit, but I need to do some additional testing to be sure.

Reply:
> The difference between ';' and semi here is important, as far as I remember.

What's the difference? They are completely equal tokens, excluding the way of designation.

Reply:
`semi` is not a token, it's a full blown rule.

Reply:
Sorry, it's my fault.